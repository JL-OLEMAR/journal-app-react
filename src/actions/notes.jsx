import { collection, deleteDoc, addDoc, updateDoc, doc } from 'firebase/firestore'
import Swal from 'sweetalert2/dist/sweetalert2.js'
import 'sweetalert2/dist/sweetalert2.css'

import { db } from '../firebase/firebase-config.jsx'
import { fileUpload } from '../helpers/fileUpload.jsx'
import { loadNotes } from '../helpers/loadNotes.jsx'
import { types } from '../types/types.jsx'

// Note active(selected)
export const activeNote = (id, note) => ({
  type: types.notesActive,
  payload: { id, ...note }
})

// Add new note
export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: { id, ...note }
})

// Add new note to firestore
export const startNewNote = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth

    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime() // in miliseconds
    }

    try {
      const doc = await addDoc(collection(db, `${uid}/journal/notes`), newNote)

      dispatch(addNewNote(doc.id, newNote))
      dispatch(activeNote(doc.id, newNote))
    } catch (error) {
      console.log(error)
    }
  }
}

// Load notes
export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes
})

// Load notes Firebase
export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    try {
      const notes = await (await loadNotes(uid))

      // sort notes by date desc
      const orderedNotes = notes.sort((a, b) => b.date - a.date)

      dispatch(setNotes(orderedNotes))
    } catch (error) {
      console.log(error)
    }
  }
}

// Reload a note with your fields save or updated
export const refreshNote = (id, note) => ({
  type: types.notesUpdated,
  payload: {
    id,
    note: { id, ...note }
  }
})

// Save or update a note to Firebase
export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth

    !note.url && delete note.url

    try {
      const noteToFirestore = {
        ...note,
        date: new Date().getTime()
      }

      delete noteToFirestore.id
      await updateDoc(doc(db, `${uid}/journal/notes`, `${note.id}`), noteToFirestore)

      dispatch(refreshNote(note.id, noteToFirestore))
      Swal.fire('Note Saved', note.title, 'success')
    } catch (error) {
      console.log(error)
    }
  }
}

// Upload image Firebase
export const startUploading = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes

    Swal.fire({
      title: 'Uploading...',
      text: 'Please wait...',
      allowOutsideClick: false,
      didOpen: () => Swal.showLoading()
    })

    const fileUrl = await fileUpload(file)

    // Set the name of the url to the selected note
    activeNote.url = fileUrl
    dispatch(startSaveNote(activeNote))
    Swal.close()
  }
}

// Delete note Firebase
export const startDeleting = (id) => {
  return async (dispatch, getState) => {
    try {
      const { uid, name } = getState().auth

      await deleteDoc(doc(db, `${uid}/journal/notes`, `${id}`))
      Swal.fire({
        title: `${name}, Are you sure?`,
        text: "You won't be able to revert this!",
        icon: 'warning',
        showCancelButton: true,
        confirmButtonColor: '#d33',
        cancelButtonColor: '#3085d6',
        confirmButtonText: 'Yes, delete it!'
      }).then((result) => {
        if (result.isConfirmed) {
          Swal.fire(
            'Deleted!',
            'Your note has been deleted.',
            'success'
          )
          dispatch(deleteNote(id))
        }
      })
    } catch (error) {
      console.log(error)
    }
  }
}

// Delete note
export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: id
})

// Clear note
export const noteLogout = () => ({
  type: types.notesLogoutCleaning
})
