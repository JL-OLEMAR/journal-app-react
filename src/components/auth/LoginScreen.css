.social-networks {
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: center;
  padding-bottom: 20px;
  padding-top: 10px;
  width: 100%;

  & .btn-google {
    background-color: #4285f4;
    border-radius: 2px;
    box-shadow: 0 3px 4px 0 #00000025;
    cursor: pointer;
    height: 42px;
    width: 100%;
    transition: box-shadow 0.3s ease;

    & .google-icon-wrapper {
      background-color: #fff;
      border-radius: 2px;
      height: 40px;
      margin-left: 1px;
      margin-top: 1px;
      position: absolute;
      width: 40px;

      & .google-icon {
        height: 18px;
        margin-left: 11px;
        margin-top: 11px;
        position: absolute;
        width: 18px;
      }
    }

    & .btn-text {
      color: #fff;
      float: right;
      font-size: 14px;
      letter-spacing: 0.2px;
      margin: 11px 40px 0 0;
    }

    &:hover {
      box-shadow: 0 0 6px #4285f4;
    }

    &:active {
      background: #1669f2;
    }
  }
}
