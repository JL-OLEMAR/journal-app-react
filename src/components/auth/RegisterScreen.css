.alert-error {
  background-color: red;
  border-radius: 5px;
  color: #fff;
  display: flex;
  justify-content: center;
  margin-bottom: 10px;
  padding: 5px;
}
