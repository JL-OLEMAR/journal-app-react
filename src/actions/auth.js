import Swal from 'sweetalert2'
import { firebase, googleAuthProvider } from '../firebase/firebase-config'
import { types } from '../types/types'
import { noteLogout } from './notes'
import { uiFinishLoading, uiStartLoading } from './ui'

export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {
    dispatch(uiStartLoading())

    firebase.auth().signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName))
        dispatch(uiFinishLoading())
      })
      .catch(e => {
        console.log(e)
        dispatch(uiFinishLoading())
        Swal.fire('Error', e.message, 'error')
      })
  }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    firebase.auth().createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name })
        dispatch(
          login(user.uid, user.displayName)
        )
      })
      .catch(e => {
        console.log(e)
        Swal.fire('Error', e.message, 'error')
      })
  }
}

export const startGoogleLogin = () => {
  return (dispatch) => {
    firebase.auth().signInWithPopup(googleAuthProvider)
      .then(({ user }) => {
        dispatch(
          login(user.uid, user.displayName)
        )
      })
  }
}

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName
  }
})

export const startLogout = () => {
  return async (dispatch) => {
    await firebase.auth().signOut()

    dispatch(logout())
    dispatch(noteLogout())
  }
}

export const logout = () => ({
  type: types.logout
})
