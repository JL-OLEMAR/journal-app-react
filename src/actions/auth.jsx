import Swal from 'sweetalert2'
import {
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  signInWithPopup,
  signOut
} from 'firebase/auth'

import { types } from '../types/types.jsx'
import { auth, googleAuthProvider } from '../firebase/firebase-config.jsx'

import { noteLogout } from './notes.jsx'
import { uiFinishLoading, uiStartLoading } from './ui.jsx'

// Login
export const login = (uid, displayName) => ({
  type: types.login,
  payload: { uid, displayName }
})

export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {
    dispatch(uiStartLoading())

    return signInWithEmailAndPassword(auth, email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName))
        dispatch(uiFinishLoading())
      })
      .catch(e => {
        console.log(e)
        dispatch(uiFinishLoading())
        Swal.fire('Error', e.message, 'error')
      })
  }
}

// Google Login
export const startGoogleLogin = () => {
  return (dispatch) => {
    signInWithPopup(auth, googleAuthProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName))
      })
  }
}

// Register
export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name })
        dispatch(login(user.uid, user.displayName))
      })
      .catch(e => {
        console.log(e)
        Swal.fire('Error', e.message, 'error')
      })
  }
}

// Logout
export const logout = () => ({
  type: types.logout
})

export const startLogout = () => {
  return async (dispatch) => {
    await signOut(auth)

    dispatch(logout())
    dispatch(noteLogout())
  }
}
