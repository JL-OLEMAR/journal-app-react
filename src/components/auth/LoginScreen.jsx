import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'

import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth.jsx'
import { useForm } from '../../hooks/useForm.jsx'
import './LoginScreen.css'

export const LoginScreen = () => {
  const dispatch = useDispatch()
  const { loading } = useSelector(state => state.ui)

  const [formValues, handleInputChange] = useForm({
    email: 'test1@test.com',
    password: '123456'
  })

  const { email, password } = formValues

  const handleLogin = (e) => {
    e.preventDefault()
    dispatch(startLoginEmailPassword(email, password))
  }

  const handleGoogleLogin = () => dispatch(startGoogleLogin())

  return (
    <>
      <h3 className='auth-title'>Login</h3>
      <form
        className='animate__animated animate__fadeIn animate__faster'
        onSubmit={handleLogin}
      >
        <input
          autoComplete='off'
          className='auth-input'
          name='email'
          placeholder='Email'
          type='text'
          value={email}
          onChange={handleInputChange}
        />

        <input
          className='auth-input'
          name='password'
          placeholder='Password'
          type='password'
          value={password}
          onChange={handleInputChange}
        />

        <button
          className='btn btn-primary btn-block auth-btn'
          disabled={loading}
          type='submit'
        >
          Login
        </button>

        <div className='social-networks'>
          <p>or</p>
          <div
            className='btn-google'
            onClick={handleGoogleLogin}
          >
            <div className='google-icon-wrapper'>
              <img
                alt='google button'
                className='google-icon'
                src='https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg'
              />
            </div>
            <p className='btn-text'>
              <b>Sign in with google</b>
            </p>
          </div>
        </div>

        <Link
          className='link conf-center'
          to='/auth/register'
        >
          Create new account
        </Link>
      </form>
    </>
  )
}
