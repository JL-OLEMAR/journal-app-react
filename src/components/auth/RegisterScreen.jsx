import { Link } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import validator from 'validator'

import { startRegisterWithEmailPasswordName } from '../../actions/auth.jsx'
import { removeError, setError } from '../../actions/ui.jsx'
import { useForm } from '../../hooks/useForm.jsx'

export const RegisterScreen = () => {
  const dispatch = useDispatch()
  const { msgError } = useSelector(state => state.ui)

  const [formValues, handleInputChange] = useForm({
    name: '',
    email: '',
    password: '',
    password2: ''
  })

  const { name, email, password, password2 } = formValues

  const handleRegister = (e) => {
    e.preventDefault()

    if (isFormValid()) {
      dispatch(startRegisterWithEmailPasswordName(name, email, password))
    }
  }

  const isFormValid = () => {
    if (name.trim().length === 0) {
      dispatch(setError('Name is required!'))

      return false
    } else if (!validator.isEmail(email)) {
      dispatch(setError('Email is not valid!'))

      return false
    } else if (password !== password2 || password.length < 5) {
      dispatch(setError('Password should be at least 6 characters and match each other'))

      return false
    }
    dispatch(removeError())

    return true
  }

  return (
    <>
      <h3 className='auth-title'>Register</h3>
      <form
        className='animate__animated animate__fadeIn animate__faster'
        onSubmit={handleRegister}
      >
        {msgError && <div className='alert-error config-center'>{msgError}</div>}

        <input
          autoComplete='off'
          className='auth-input'
          name='name'
          placeholder='Name'
          type='text'
          value={name}
          onChange={handleInputChange}
        />

        <input
          autoComplete='off'
          className='auth-input'
          name='email'
          placeholder='Email'
          type='text'
          value={email}
          onChange={handleInputChange}
        />

        <input
          className='auth-input'
          name='password'
          placeholder='Password'
          type='password'
          value={password}
          onChange={handleInputChange}
        />

        <input
          className='auth-input'
          name='password2'
          placeholder='Confirm password'
          type='password'
          value={password2}
          onChange={handleInputChange}
        />

        <button
          className='btn btn-primary btn-block mb-5 auth-btn'
          type='submit'
        >
          Register
        </button>

        <Link
          className='link config-center'
          to='/auth/login'
        >
          Already registered?
        </Link>
      </form>
    </>
  )
}
