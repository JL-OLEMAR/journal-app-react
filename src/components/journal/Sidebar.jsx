import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'

import { startLogout } from '../../actions/auth.jsx'
import { startNewNote } from '../../actions/notes.jsx'

import { JournalEntries } from './JournalEntries.jsx'

export const Sidebar = () => {
  const dispatch = useDispatch()
  const { name } = useSelector(state => state.auth)

  const handleLogout = () => {
    dispatch(startLogout())
  }

  const handleAddNew = () => {
    dispatch(startNewNote())
  }

  const firstletterUpperCase = name && name.charAt(0).toUpperCase() + name.slice(1)

  return (
    <aside className='journal-sidebar'>
      <div className='navbar'>
        <h3 className='mt-2'>
          <i className='far fa-moon' />
          <span>{firstletterUpperCase}&apos;s Journal</span>
        </h3>

        <button className='btn btn-danger logout-navbar mt-2' onClick={handleLogout}>
          Logout
        </button>
      </div>

      <div className='new-entry-container mt-2 config-center'>
        <Link className='ancla-note' to='/#note'>
          <button className='btn btn-new-entry' onClick={handleAddNew}>
            <i className='far fa-calendar-plus fa-5x' />
            <p className='mt-2'>New entry</p>
          </button>
        </Link>
      </div>
      <JournalEntries />
    </aside>
  )
}
