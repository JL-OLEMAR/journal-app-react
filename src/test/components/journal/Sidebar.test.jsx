/**
 * @jest-environment node
 */

import { beforeEach, describe, expect, test, vi } from 'vitest'
import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import { mount } from 'enzyme'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import { startLogout } from '../../../actions/auth.jsx'
import { startNewNote } from '../../../actions/notes.jsx'
import { Sidebar } from '../../../components/journal/Sidebar.jsx'
import 'jsdom-global/register' // → For mount a component

vi.mock('../../../actions/auth.jsx', () => ({
  startLogout: vi.fn()
}))

vi.mock('../../../actions/notes.jsx', () => ({
  startNewNote: vi.fn()
}))

const middlewares = [thunk]
const mockStore = configureStore(middlewares)
const initState = { // → See in Redux DevTools
  auth: {
    uid: '123',
    name: 'John Doe'
  },
  ui: {
    loading: false,
    msgError: null
  },
  notes: {
    active: null,
    notes: []
  }
}
let store = mockStore(initState) // → mock store

store.dispatch = vi.fn() // → Mock dispatch

const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <Sidebar />
    </MemoryRouter>
  </Provider>
)

describe('Tests <Sidebar />', () => {
  beforeEach(() => {
    store = mockStore(initState)
    vi.clearAllMocks()
  })

  test('should show correctly \'snapshot\'', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('should call the startLogout action', () => {
    wrapper.find('button.btn.btn-danger').prop('onClick')()
    expect(startLogout).toHaveBeenCalled()
  })

  test('should call the startNewNote action', () => {
    wrapper.find('button.btn.btn-new-entry').prop('onClick')()
    expect(startNewNote).toHaveBeenCalled()
  })
})
