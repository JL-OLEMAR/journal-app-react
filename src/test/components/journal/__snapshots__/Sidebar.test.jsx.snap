// Vitest Snapshot v1

exports[`Tests <Sidebar /> > should show correctly 'snapshot' 1`] = `
<aside
  className="journal-sidebar"
>
  <div
    className="navbar"
  >
    <h3
      className="mt-2"
    >
      <i
        className="far fa-moon"
      />
      <span>
        John Doe's Journal
      </span>
    </h3>
    <button
      className="btn btn-danger logout-navbar mt-2"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="new-entry-container mt-2 config-center"
  >
    <a
      className="ancla-note"
      href="/#note"
      onClick={[Function]}
    >
      <button
        className="btn btn-new-entry"
        onClick={[Function]}
      >
        <i
          className="far fa-calendar-plus fa-5x"
        />
        <p
          className="mt-2"
        >
          New entry
        </p>
      </button>
    </a>
  </div>
  <div
    className="journal-entries"
  />
</aside>
`;
